src/NUnitFramework/framework/Internal/Results/TestResult.cs

```diff
@@ -498,24 +498,15 @@ public void RecordException(Exception ex)
             if (ex is NUnitException || ex is TargetInvocationException)
                 ex = ex.InnerException;
 
-            if (ex is AssertionException)
-            {
-                AssertionResults.Add(new AssertionResult(AssertionStatus.Failed, ex.Message, StackFilter.DefaultFilter.Filter(ex.StackTrace)));
-            }
-
             if (ex is ResultStateException)
-            {
-                string message = ex is MultipleAssertException
-                    ? CreateLegacyFailureMessage()
-                    : ex.Message;
-
-                string stackTrace = StackFilter.DefaultFilter.Filter(ex.StackTrace);
-
-                SetResult(((ResultStateException)ex).ResultState, message, stackTrace);
-            }
+                SetResult(
+                    ((ResultStateException)ex).ResultState,
```

Comment:
This looks unnecessarily complex. I'd suggest either changing this logic into `as`->`!= null` (to avoid the code duplication), or even better, rely on the new switch with pattern matching.

I know this is probably out of scope for this PR though.